text/0000-iri-v181-transaction-and-bundle/0000-iri-v181-transaction-and-bundle.md

```diff
@@ -0,0 +1,489 @@
+subset of the transaction fields.
+
+| Name          | Size      |
+| ------------- | --------- |
+| address       | 243 trits |
+| value         | 81 trits  |
+| obsolete_tag  | 81 trits  |
+| timestamp     | 27 trits  |
+| current_index | 27 trits  |
+| last_index    | 27 trits  |
+
+The bundle hash is generated with a sponge by iterating through the bundle, from `0` to `last_index`, absorbing the
+bundle essence of each transaction and eventually squeezing the bundle hash from the sponge.
+
+Pseudocode:
+
+```
+calculate_hash(bundle)
+| sponge = Sponge(Kerl)
+|
+| for transaction in bundle
+| | sponge.absorb(transaction.essence())
+|
+| return sponge.squeeze()
+```
+
+#### Finalise
+
+*Client side operation.*
+
+Finalising a bundle means computing the bundle hash, verifying that it matches the security requirement and setting it
+to all the transactions of the bundle. After finalisation, transactions of a bundle are ready to be safely attached to
+the tangle.
+
+Pseudocode:
+
+```
+finalise(bundle)
+| hash = bundleHash(bundle)
+|
+| while hash.normalise().find('M')
+| | bundle.at(0).obsolete_tag++
+| | hash = bundleHash(bundle)
+|
+| for transaction in bundle
+| | transaction.setBundleHash(hash)
+```
+
+*Security requirement: due to the implementation of the signature process, the normalised bundle hash can't contain a
+`M` or `13` because it could expose a significant part of the private key, weakening the signature. The bundle hash is
+then repetitively generated with a slight modification until its normalisation doesn't contain a `M`.*
+
+#### Sign
+
+*Client side operation.*
+
+Signing a bundle allow you to prove that you are the owner of the address you are trying to move funds from. With no
+signature or a bad signature, the bundle won't be considered valid and the funds won't be moved. Only the owner of the
+right seed is able to generate the right signature for this address.
+
+Pseudocode:
+
+```
+sign(bundle, seed, inputs)
+| current_index = 0
+|
+| for transaction in bundle
+| | if transaction.value < 0
+| | | if transaction.current_index >= current_index
+| | | | input = inputs[transaction.address]
+| | | | fragments = sign(seed, input.index, input.security, transaction.bundle)
+| | | | for fragment in fragments
+| | | | | bundle.at(current_index).signature = fragment
+| | | | | current_index = current_index + 1
+| | else
+| | | current_index = current_index + 1
+```
+
+*Since signature size depends on the security level, a single signature can spread out to up to 3 transactions.
+`inputs` is an object that contain all unused addresses of a seed with a sufficient balance.*
+
+#### Proof of Work
+
+*Client side operation.*
+
+Proof of Work (PoW) allows your transactions to be accepted by the network. On the IOTA network, PoW is only a rate
+control mechanism. Doing PoW on a bundle means doing PoW on each of its transactions and setting trunks and branch
```

Comment:
```suggestion
PoW is meant to prevent Sybil attack. Doing PoW on a bundle means doing PoW on each of its transactions and setting trunks and branch
```

Reply:
PoW is a rate control mechanism, not an anti-Sybil protection. Mana is.